rclcpp/src/rclcpp/executors/static_executor_waitable.cpp

```diff
@@ -0,0 +1,264 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/executors/static_single_threaded_executor.hpp"
+#include "rclcpp/executors/static_executor_waitable.hpp"
+
+using rclcpp::executors::StaticExecutorWaitable;
+
+StaticExecutorWaitable::~StaticExecutorWaitable()
+{
+  // Disassociate all nodes
+  for (auto & weak_node : weak_nodes_) {
+    auto node = weak_node.lock();
+    if (node) {
+      std::atomic_bool & has_executor = node->get_associated_with_executor_atomic();
+      has_executor.store(false);
+    }
+  }
+  exec_list_.clear();
+  weak_nodes_.clear();
+  guard_conditions_.clear();
+}
+
+void
+StaticExecutorWaitable::init(
+       rcl_wait_set_t* p_wait_set,
+       memory_strategy::MemoryStrategy::SharedPtr& memory_strategy,
+       StaticExecutorWaitable::SharedPtr this_shared_ptr)
+{
+  // Empty initialize executable list
+  executor::ExecutableList exec_list_();
+  // Get executor's wait_set_ pointer
+  p_wait_set_ = p_wait_set;
+  // Get executor's memory strategy ptr
+  if (memory_strategy == nullptr) {
+    throw std::runtime_error("Received NULL memory strategy in executor waitable.");
+  }
+  memory_strategy_ = memory_strategy;
+  // Get this StaticExecutorWaitable shared pointer
+  this_shared_ptr_ = this_shared_ptr;
+
+  // Get memory strategy and executable list. Prepare wait_set_
+  execute();
+}
+
+void
+StaticExecutorWaitable::execute()
+{
+  // Fill memory strategy with entities comming from weak_nodes_
+  get_memory_strategy();
+  // Fill exec_list_ with entities comming from weak_nodes_ (same as memory strategy)
+  get_executable_list();
+  // Resize the wait_set_ based on memory_strategy handles (rcl_wait_set_resize)
+  prepare_wait_set();
+}
+
+void
+StaticExecutorWaitable::get_memory_strategy()
```

Comment:
This should be renamed.

From its name I would expect that it returns the memory strategy